torch/distributed/_state_dict_utils.py

```diff
@@ -185,26 +236,152 @@ def dtensor_func(value, pg, device):
+            have the same state_dicts. Other ranks will get empty state_dicts.
+        cpu_offload_state_dict (Optional[Dict[str, Any]]): the CPU state_dict
+            that will be returned. If this is not None, this API will use
+            `copy_` to copy the GPU tensor to the tensor in this CPU state_dict.
+            This CPU state_dict must have exactly the same structure as the
+            `state_dict` the only difference is that all the tensors in this
+            CPU state_dict are on CPU memory.
+        cpu_offload_sync: (bool): flag to decide whether to call `synchronize()`
+            before this API returns.
+        type_check: (bool): check if the instance data type is a supported type
+            that can be saved by DCP.  The current supported data types are
+            torch.Tensor, DTensor, int, float, str, list, dict, None.
+
+    Returns:
+        The gathered state dictionary.
+    """
+
+    ret = _iterate_state_dict(
         state_dict,
-        lambda value, pg, device: value,
-        lambda value, pg, device: value,
-        pg=pg,
-        device=device,
+        _identity_func,
+        _identity_func,
+        _identity_func,
+        pg=None,
+        device=None,
         cpu_offload=True,
         ranks_only=ranks_only,
+        companion_obj=cpu_offload_state_dict,
+        type_check=type_check,
+    )
+    if cpu_offload_state_dict is not None and cpu_offload_sync:
+        torch.cuda.synchronize()
+    return ret
+
+
+def _create_cpu_state_dict(
+    state_dict: Dict[str, Any], pin_memory: bool = False, share_memory: bool = False
+) -> Dict[str, Any]:
+    """
+    Given a state_dict, create another state_dict with the same structure and elements.
+    However, all tensors in the returned state_dict are new tensors on CPU. These
+    tensors can be placed on pin_memory or share_memory based on the provided arguments.
+    """
+
+    if pin_memory and share_memory:
+        raise ValueError(
+            "Cannot allocate both memory on both pin_memory and share_memory"
+        )
+
+    def tensor_func(
+        obj: torch.Tensor,
+        pg: Optional[dist.ProcessGroup],
+        device: Optional[torch.device],
+        companion_obj: Any,
+    ) -> torch.Tensor:
+        if len(obj.size()) == 0:
+            return torch.tensor(0, dtype=obj.dtype)
+
+        if share_memory:
+            return torch.empty(
+                *tuple(companion_obj.size()), dtype=companion_obj.dtype
+            ).share_memory_()
+        else:
+            return torch.empty(
+                *tuple(companion_obj.size()), dtype=companion_obj.dtype
+            ).pin_memory()
+
+    ret = _iterate_state_dict(
+        state_dict,
+        _identity_func,
+        _identity_func,
+        tensor_func,
+        pg=None,
+        device=None,
+        cpu_offload=False,
+        ranks_only=tuple(),
+        companion_obj=state_dict,
+        type_check=False,
     )
+    return ret
+
+
+def _check_state_dict_similarity(
```

Comment:
Would it be difficult to always do this when we have a companion obj in `_iterate` ? That way users wouldn't need to call this method before using pinned memory, etc.

Reply:
It's not difficult but I'm not sure if it is better. If we check in the offload_cpu call, users have to do EAFP style, using a try-catch to see if there are errors. However, in this case, if the exception happens, users will need to do the mitigation and retry offload_cpu call. The error is not an actual error as it is possible that state_dict changes. If users just use LBYL style, using a if statement to check, it is simpler IMHO. No retry just a check anyway.